jac-uk/apply
# #1356 Fixes to QT timer

## What's included?
Fixes issues with the QT countdown timer.

## Who should test?
✅ Product owner
✅ Developers
✅ UTG

## How to test?
You'll need to set up a new QT. The `duration` can be configured to be quite short (e.g. 5 minutes) so these tests don't take too long 😄 

Would recommend running the following against the live or staging site as well as the preview Url.
On live / staging the issue can be observed.
On the preview url the fix can be observed.

### Test 1: Local clock is in the future

- Set your computer/tablet local time to X minutes in the future (e.g. 10:10 instead of 10:00)
- Open your web browser, navigate to apply site and start the test
- Keep the test open and observe what happens, particularly X minutes before the expected end of the test
- Is everything as expected?

_Issue being fixed:_ test closes X minutes early
[Video](https://drive.google.com/file/d/1vPx2OOye4rEp2_HUUhnGx_movnIL_W8g/view?usp=sharing)

### Test 2: Local clock is in the past

- Set your computer/tablet local time to X minutes in the past (e.g. 09:50 instead of 10:00)
- Open your web browser, navigate to apply site and start the test
- Keep the test open and observe what happens
- Is everything as expected?

_Issue being fixed:_ test stays open an additional X minutes

### Test 3: During the test the local clock is changed to be in the future

- Open your web browser, navigate to apply site and start the test
- Set your computer/tablet local time to X minutes in the future (e.g. 10:10 instead of 10:00)
- Keep the test open and observe what happens
- Is everything as expected?

_Issue being fixed:_ user can cheat by changing their local clock
[Video](https://drive.google.com/file/d/1HrWqBA7sBg86_MCpE9s9GmBGuKSPDtPx/view?usp=sharing)

### Test 4: During the test the local clock is changed to be in the past

- Open your web browser, navigate to apply site and start the test
- Set your computer/tablet local time to X minutes in the past (e.g. 09:50 instead of 10:00)
- Keep the test open and observe what happens
- Is everything as expected?

_Issue being fixed:_ user can cheat by changing their local clock
[Video](https://drive.google.com/file/d/1Ar9ua98XXP8QL9jRjlBHy2Buqx0XuKs8/view?usp=sharing)

### Test 5: Starting the test after it has expired

- Navigate to apply site and start the test 
- Answer a question then exit the test 
- Observe that the test is in the ‘Open tests’ list on the qualifying tests page 
- Wait until the test has expired (e.g. for a 10 minute test wait more than 10 minutes) 
- Is everything as expected? 
- Is it possible to start the test again?

_Issue being fixed:_ user can start the test after it has expired


## Risk - how likely is this to impact other areas?
🟠 Medium risk - this does change code that is shared with other areas

---
PREVIEW:DEVELOP
